### DISCUSSION TITLE
Extending full os module in MP

### DISCUSSION THREAD

**Discussion by Qdog31** [micropython/micropython#11833] [Other]
Hello, I have been working with MP for a couple of weeks now and was wondering if someone could help me with figuring out how to add the remaining attributes to the os module within MP? 

I currently have a compiled unix port executable and am attempting to utilize the entirety of the python os module within the MP REPL. However, I only have the following attributes:  
```py
>> dir(os)

[__'class__', __'name__', 'remove', 'sep', '__dict__', 'VfsFat', 'VfsLfs1', 'VfsLfs2', 'VfsPosix', 'chdir', 'errno', 'getcwd', 'getenv', 'ilistdir', 'listdir', 'mkdir', 'mount', 'putenv', 'rename', 'rmdir', 'stat', 'statvfs', 'system', 'umount', 'unlink', 'unsetenv', 'urandom']
````

My question: When running the same command (dir(os)) with python3, it seems like there are well over a hundred. How can I add the rest of these attributes? If this is not possible, I am interested in learning how to atleast manually add ones that I choose. For example, the os.walk() method would be useful to have. 

I have been messing around with the usercmodule example, but am super confused on how I could modify these files with the os module code.

**Comment by dlech** [comment]
> attempting to utilize the entirety of the python os module

What is the motivation for this?


> How can I add the rest of these attributes?

You can modify the [modos.c](https://github.com/micropython/micropython/blob/master/ports/unix/modos.c) file if you need to implement something in C or [os](https://github.com/micropython/micropython-lib/blob/master/python-stdlib/os/os/__init__.py) in micropython-lib if it can be implemented in Python.

**Comment by bixb922** [comment]
As @dlech points out, these functions can be implemented also in Python. For example a search for "os.walk source code" brings up some results, including the source of CPython. These results will have to be adapted to Micropython.
As to why os.walk and other functions are not implemented: what I believe is that Micropython has to run on very small microcontrollers. What functions to include is a tough decision, but I think the developers did a good job providing what's needed for the community to build upon. For example, os.listdir is present.

**Comment by Qdog31** [comment]
@dlech @bixb922 Thank you for your responses! I do not want to add the **entire** os module, but I definitely would like other common attributes such as walk and other unique ones for complex problems. Now, I think the next step is to figure out how to adapt this code for MicroPython. If anyone could help me create an example folder/file to add the os.walk functionality that would be helpful. Either way, I look forward to spending the next couple of days understanding the MP syntax.

**Comment by dlech** [comment]
The unix-ffi package in micropython-lib has a bunch of stuff implemented already: https://github.com/micropython/micropython-lib/blob/master/unix-ffi/os/os/__init__.py

**Comment by jimmo** [comment]
@Qdog31 As already pointed out we provide an example of extending the `os` module in micropython-lib.

For more documentation on extending built-in modules from Python see this recently updated documentation: https://docs.micropython.org/en/latest/library/index.html#extending-built-in-libraries-from-python

**Comment by Qdog31** [comment]
@dlech I'm currently looking at the file you shared, and it has a walk definition. However when I build the unix port, there still is not a walk attribute. How can I add this? 
@jimmo I am also reading the extending built-in modules documentation and cannot figure out how to implement it. I created an "os.py" file that has the extra method (os.walk), but do not know where to put it or if I'm doing it correct. For my os.py file I have: 
```
_path = sys.path
sys.path = () 
try: 
    from os import * 
finally: 
    sys.path = _path
   del _path 

def walk(top, topdown=True, onerror=None, followlinks=False): 
...
```
I currently placed this file in all directories having to deal with lib, because I am not sure where it goes. Does it go within /lib/micropython-lib or /lib/micropython-lib/python-stdlib, etc? Lastly, would it be easier to add modules this way or to freeze them? I am still looking into the differences.

**Reply by dlech** [reply]
You can `print(sys.path)` to see what the default path is for searching for modules. It should be enough to just put `os.py` in the working directory where you are running `micropython` from.
